iceoryx_posh/include/iceoryx_posh/popo/active_call_set.hpp

```diff
@@ -0,0 +1,197 @@
+    /// @note This method can be called from any thread concurrently without any restrictions!
+    /// @tparam[in] T type of the class which will signal the event
+    /// @param[in] eventOrigin the object which will signal the event (the origin)
+    /// @param[in] eventType enum required to specify the type of event inside of eventOrigin
+    /// @param[in] eventCallback callback which will be executed concurrently when the event occurs
+    /// @return If an error occurs the enum which describes the error.
+    template <typename T, typename EventType, typename = std::enable_if_t<std::is_enum<EventType>::value>>
+    cxx::expected<ActiveCallSetError>
+    attachEvent(T& eventOrigin, const EventType eventType, CallbackRef_t<T> eventCallback) noexcept;
+
+    /// @brief Detaches an event. Hereby, the event is defined as a class T, the eventOrigin and
+    ///        the eventType with further specifies the event inside of eventOrigin
+    /// @note This method can be called from any thread concurrently without any restrictions!
+    /// @tparam[in] T type of the class which will signal the event
+    /// @param[in] eventOrigin the object which will signal the event (the origin)
+    /// @param[in] eventType enum required to specify the type of event inside of eventOrigin
+    template <typename T, typename EventType, typename = std::enable_if_t<std::is_enum<EventType>::value>>
+    void detachEvent(T& eventOrigin, const EventType eventType) noexcept;
+
+    /// @brief Detaches an event. Hereby, the event is defined as a class T, the eventOrigin.
+    /// @note This method can be called from any thread concurrently without any restrictions!
+    /// @tparam[in] T type of the class which will signal the event
+    /// @param[in] eventType enum required to specify the type of event inside of eventOrigin
+    template <typename T>
+    void detachEvent(T& eventOrigin) noexcept;
+
+    /// @brief Returns the capacity of the ActiveCallSet
+    /// @return capacity of the ActiveCallSet
+    static constexpr uint64_t capacity() noexcept;
+
+    /// @brief Returns the size of the ActiveCallSet
+    /// @return size of the ActiveCallSet
+    uint64_t size() const noexcept;
+
+  protected:
+    ActiveCallSet(EventVariableData* eventVariable) noexcept;
+
+  private:
+    class Event_t;
+
+    void threadLoop() noexcept;
+    cxx::expected<uint64_t, ActiveCallSetError>
+    addEvent(void* const origin,
+             const uint64_t eventType,
+             const uint64_t eventTypeHash,
+             CallbackRef_t<void> callback,
+             TranslationCallbackRef_t translationCallback,
+             const cxx::MethodCallback<void, uint64_t> invalidationCallback) noexcept;
+    void removeEvent(void* const origin, const uint64_t eventType, const uint64_t eventTypeHash) noexcept;
+
+    void removeTrigger(const uint64_t index) noexcept;
+
+  private:
+    enum class NoEnumUsed
+    {
+        PLACEHOLDER = 0
+    };
+
+    class Event_t
```

Comment:
This is a nested class and is only used by the ActiveCallSet.

You cannot create a definition for a nested class in a separate header and I do not want that this class is used outside of the ActiveCallSet.

Reply:
@elBoberido In my opinion I tested the full class via the external behavior of the `active_call_set`. Actually, it helped me quiet a bit to review the internal class and write tests for the active call set otherwise you rely on the unit test of the `Event_t` and think you do not have to retest the behavior in the `active_call_set`.
But if you think of them as one union, that the one cannot exist without the other I think the test quality increases somehow ...